lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1760,6 +1760,40 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
   case OpGetKernelNDrangeSubGroupCount:
     return mapValue(
         BV, transSGSizeQueryBI(static_cast<SPIRVInstruction *>(BV), BB));
+  case OpFPGARegINTEL: {
+    IRBuilder<> Builder(BB);
+
+    SPIRVFPGARegINTELInstBase *BC =
+        static_cast<SPIRVFPGARegINTELInstBase *>(BV);
+
+    Type *Int8PtrTyPrivate = Type::getInt8PtrTy(*Context, SPIRAS_Private);
+    IntegerType *Int32Ty = Type::getInt32Ty(*Context);
+
+    Value *UndefInt8Ptr = UndefValue::get(Int8PtrTyPrivate);
+    Value *UndefInt32 = UndefValue::get(Int32Ty);
+
+    StringRef AnnotStr("__builtin_intel_fpga_reg");
+    Constant *GS = BB->getModule()->getNamedGlobal("__reg");
+    if (!GS)
+      GS = Builder.CreateGlobalStringPtr(AnnotStr, "__reg");
+
+    Function *AnnotationFn;
+    if (auto RetIntTy = dyn_cast<IntegerType>(transType(BC->getType()))) {
+      AnnotationFn =
+          llvm::Intrinsic::getDeclaration(M, Intrinsic::annotation, RetIntTy);
+    } else if (auto RetPtrTy =
+                   dyn_cast<PointerType>(transType(BC->getType()))) {
+      AnnotationFn = llvm::Intrinsic::getDeclaration(
+          M, Intrinsic::ptr_annotation, RetPtrTy);
+    }
+
+    llvm::Value *Args[] = {transValue(BC->getOperand(0), F, BB),
+                           Builder.CreateBitCast(GS, Int8PtrTyPrivate),
```

Comment:
Are you sure you need this bitcast? CreateGlobalStringPtr returns i8*.

Reply:
This has to do with the address space coherence. `CreateGlobalStringPtr` returns the global pointer, so `CreateBitCast` casts the pointer to `Int8PtrPrivate`, which is in turn declared with `SPIRAS_Private` argument. Additionally, please see `Int8PtrTyPrivate` declaration at line 1772.

Reply:
If you want to cast between address spaces, then you need IRBuilder.CreateAddrSpaceCast.
Anyway, I don't think you should do this in the first place. Global and private address spaces do not overlap, and an addrspacecast between them would be is invalid. Can the annotate intrinsic take a pointer in global address space?

Reply:
After consideration, I don't see why the intrinsic should not accept a global AS pointer. Will fix despite the natural "it works though!" :)

Reply:
According to Intrinsics.td, annotation intrinsic only supports pointers in (0) address space, so this cast is required.

Reply:
>  `CreateGlobalStringPtr`  returns the global pointer

Not necessarily.

Actually `CreateGlobalStringPtr` takes address space as a parameter. See here https://github.com/llvm/llvm-project/blob/52c39396151978ca946e2a80d9118c8672bace14/llvm/include/llvm/IR/IRBuilder.h#L1861-L1862
Moreover its default value is 0.